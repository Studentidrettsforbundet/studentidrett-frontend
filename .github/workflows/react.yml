on:
  pull_request:
    branches:
      - dev

name: Frontend build and push to Azure

jobs:
    build-and-deploy:
        env:
          DJANGO_SECRET_KEY: ${{ secrets.DJANGO_SECRET_KEY }}
          IMAGE_VERSION: ${{ github.actor }}
          ACR: ${{ secrets.REGISTRY_LOGIN_SERVER }}

        runs-on: ubuntu-latest
        steps:
        - name: 'Checkout GitHub Action'
          uses: actions/checkout@master

        - name: 'Login via Azure CLI'
          uses: azure/login@v1
          with:
            creds: ${{ secrets.AZURE_CREDENTIALS }}

        - name: 'Build image'
          run: |
            docker build . -t ${{ secrets.REGISTRY_LOGIN_SERVER }}/studentidrett-frontend:dev
            
        - name: 'Push image to ACR'
          uses: azure/docker-login@v1
          with:
            login-server: ${{ secrets.REGISTRY_LOGIN_SERVER }}
            username: ${{ secrets.REGISTRY_USERNAME }}
            password: ${{ secrets.REGISTRY_PASSWORD }}
        - run: |
            docker push ${{ secrets.REGISTRY_LOGIN_SERVER }}/studentidrett-frontend:dev
        - name: Azure logout
          run: |
            az logout
