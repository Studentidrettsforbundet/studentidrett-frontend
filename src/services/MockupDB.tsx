import React from "react"

///// Database example setup
export interface Sport {
    type: 'SPORT';
    Name: string;
    Info: string;
}

export interface Region {
    type: 'REGION';
    Name: string;
    Sports: Array<Sport>;
}

export interface City {
    type: 'CITY';
    Name: string;
    Clubs: Array<string>;
}

export interface Club {
    type: 'CLUB';
    Name: string;
    Info: string;
}

export interface Team {
    type: 'TEAM'
    Name: string;
    Text: string;
    Img: string;
}

export interface IMockupDB {
    Regions: Array<Region>;
    Sports: Array<Sport>;
    Clubs: Array<Club>;
    Teams: Array<Team>;
}

const TrondheimClubs: Array<string> = ["NTNUI", "Janus IL"];
const SteinkjerClubs: Array<string> = ["Steinkjer IL"];
const OsloClubs: Array<string> = ["OsloMet IL"];

const NTNUI: Club = {
    type: 'CLUB',
    Name: 'NTNUI',
    Info: 'NTNUI is a studentclub',
}

const JanusIL: Club = {
    type: 'CLUB',
    Name: 'Janus IL',
    Info: 'Janus IL is .. ',
}

const Trondheim: City = {
    type: "CITY",
    Name: "Trondheim",
    Clubs: TrondheimClubs,
};

const Steinkjer: City = {
    type: "CITY",
    Name: "Steinkjer",
    Clubs: SteinkjerClubs,
};

const Oslo: City = {
    type: "CITY",
    Name: "Oslo",
    Clubs: OsloClubs,
}

const TrondelagCities: Array<City> = [Trondheim, Steinkjer];
const OstfoldCities: Array<City> = [Oslo];

const Fotball: Sport = {
    type: "SPORT",
    Name: "Fotball",
    Info: 'football is nice' ,
};

const Handball: Sport = {
    type: "SPORT",
    Name: "Handball",
    Info: 'handball is ball' ,
};

const Volleyball: Sport = {
    type: "SPORT",
    Name: "Volleyball",
    Info: 'Volleyball is super',
}

const TrondelagSports: Array<Sport> = [Fotball, Handball];
const OstfoldSports: Array<Sport> = [Volleyball, Handball];

var Trondelag: Region = {
    type: "REGION",
    Name: "Trondelag",
    Sports: TrondelagSports,
}

var Ostfold: Region = {
    type: "REGION",
    Name: "Ostfold",
    Sports: OstfoldSports,
}

var DefaultCity: City = {
    type: "CITY",
    Name: "defaultCity",
    Clubs: TrondheimClubs.concat(SteinkjerClubs.concat(OsloClubs)),
}

const team1: Team = {
    type: "TEAM",
    Name: 'team 1',
    Text: ' text for team 1',
    Img : 'sunflower',
}

const team2: Team = {
    type: "TEAM",
    Name: 'team 2',
    Text: ' text for team 2',
    Img: 'japanese',
}

const team3: Team = {
    type: "TEAM",
    Name: 'team 3',
    Text: ' text for team 3',
    Img: 'sunflower',
}

let Regions: Array<Region> = [Trondelag, Ostfold];
let Sports: Array<Sport> = [Fotball, Handball, Volleyball];
let Clubs: Array<Club> = [NTNUI, JanusIL]
let Teams: Array<Team> = [team1, team2, team3]

export var MockupDB: IMockupDB = { Regions: Regions, Sports: Sports, Clubs: Clubs, Teams: Teams };


//let ThisProps: Props;

// DATABASE example setup FIN

export default MockupDB;