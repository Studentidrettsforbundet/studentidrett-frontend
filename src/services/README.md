All api methods are put in a folder called services. This removes business logic of an API call from the components, and so that any component can easily reference a service that it needs.
