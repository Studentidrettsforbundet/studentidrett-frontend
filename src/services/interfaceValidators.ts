export const instanceOfRegion = (object: any) => {
    return 'id' in object && 'name' in object && Object.keys(object).length === 2;
};

export const instanceOfTeam = (object: any): boolean => {
    return (
        'id' in object &&
        'name' in object &&
        'full_capacity' in object &&
        'short_description' in object &&
        'long_description' in object &&
        'tryouts' in object &&
        'registration_open' in object &&
        'group' in object &&
        Object.keys(object).length === 8
    );
};

export const instanceOfGroup = (object: any): boolean => {
    return (
        'id' in object &&
        'name' in object &&
        'description' in object &&
        'cover_photo' in object &&
        'sports' in object &&
        'club' in object &&
        'city' in object &&
        'contact_email' in object &&
        Object.keys(object).length === 8
    );
};

export const instanceOfSport = (object: any): boolean => {
    return (
        'id' in object &&
        'name' in object &&
        Object.keys(object).length === 2
    );
};

export const instanceOfClub = (object: any): boolean => {
    return (
        'id' in object &&
        'city' in object &&
        'name' in object &&
        'description' in object &&
        'contact_email' in object &&
        'membership_fee' in object &&
        'register_info' in object &&
        Object.keys(object).length === 7
    );
};

export const instanceOfCity = (object: any): boolean => {
    return (
        'id' in object &&
        'name' in object &&
        'region' in object &&
        'clubs' in object &&
        Object.keys(object).length === 4
    );
};
