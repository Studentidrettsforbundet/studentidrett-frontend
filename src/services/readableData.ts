export const getTeamGender = (gender?: string): string | null => {
    if (gender === 'M') {
        return 'Herre';
    } else if (gender === 'F') {
        return 'Dame';
    } else if (gender === 'MX') {
        return 'Mix';
    } else if (gender === 'A') {
        return 'Alle';
    } else {
        return null;
    }
};

export const getTeamSkill = (skill?: string): string | null => {
    if (skill === 'NONE') {
        return 'Ingen forkunnskaper krevd';
    } else if (skill === 'LOW') {
        return 'Lavt';
    } else if (skill === 'MED') {
        return 'Medium';
    } else if (skill === 'HI') {
        return 'HÃ¸yt';
    } else if (skill === 'PRO') {
        return 'Proff';
    } else {
        return null;
    }
};

export const getTeamAvailability = (availability?: string): string | null => {
    if (availability === 'OP') {
        return 'Ã…pent';
    } else if (availability === 'CL') {
        return 'Lukket';
    } else if (availability === 'FU') {
        return 'Fullt';
    } else if (availability === 'TO') {
        return 'Opptak';
    } else {
        return null;
    }
};