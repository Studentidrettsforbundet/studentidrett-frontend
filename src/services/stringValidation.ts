/**
 * This method performs all input validation: normalization, whitelisting, and URI encoding
 * @param inputString a dangerous string
 * @returns a "safe" string encoded on URI format
 */
export const makeInputSafe = (inputString: string, maxLength: number): string => {
    const normalized = stringNormalization(inputString);
    const trimmed = trimWhiteSpaces(normalized);
    const encoded = urlEncoding(trimmed);
    const lowercase = encoded.toLocaleLowerCase();

    return lowercase;
};

/**
 * This method checks if a string is within a certain length, and if it satisfies the regex.
 * @param inputString
 * @param maxLength
 */
export const validateString = (inputString: string, maxLength: number): [boolean, string] => {
    const lengthCheck = inputString.length <= maxLength;
    const regexCheck = stringWhitelisting(inputString);
    var feedbackString = '';

    if (!lengthCheck) {
        feedbackString += 'Too Long ';
    }

    if (!regexCheck) {
        feedbackString += 'Illegal Characters';
    }

    return [lengthCheck && regexCheck, feedbackString];
};

/**
 * This method removes leading and trailing whitespaces, and removes duplicate whitespaces.
 * @param inputString string with potentially many whitespaces
 * @returns string with no leading or trailing whitespaces, and only one following whitespace
 */
export const trimWhiteSpaces = (inputString: string): string => {
    const removeExternalSpaces = inputString.trim();
    return removeExternalSpaces.replace(/\s\s+/g, ' ');
};

/**
 * This method performs normalization on the input string into NFC.
 * @param inputString
 */
export const stringNormalization = (inputString: string): string => {
    return inputString.normalize('NFC');
};

/**
 * This method whitelists the inputstring. It allows all letters A-Å, Ä, Ö, and(, . - _ ')
 * @param inputString
 * @returns a boolean indicating if the string passed the whitelisting
 */
export const stringWhitelisting = (inputString: string): boolean => {
    // prettier-ignore
    var regex = new RegExp("^[a-zA-Z0-9&ÆæØøÅåÄäÖö_.,'\\-\\ ]+$");

    return regex.test(inputString);
};

/**
 * This method returns an URIComponent encoded string, which is safe for GET/PUT/POST request.
 * @param inputString The user inserted data
 * @returns URI safe encoding
 */
export const urlEncoding = (inputString: string): string => {
    return encodeURIComponent(inputString);
};
