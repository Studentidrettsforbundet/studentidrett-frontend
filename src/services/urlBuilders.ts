import { BASE_URL, cardType } from '../constants';
import { makeInputSafe } from './stringValidation';

const MAXLENGTH_SEARCH = 50;

export const urlBuilderSimpleSearch = (cardType: cardType, searchString: string): string => {
    const safeString = makeInputSafe(searchString, MAXLENGTH_SEARCH);
    return BASE_URL + '/' + cardType + '/' + safeString;
};

export const urlBuilderFetchData = (cardType: cardType): string => {
    const safeString = makeInputSafe('', MAXLENGTH_SEARCH);
    return BASE_URL + '/' + cardType + '/' + safeString;
};
