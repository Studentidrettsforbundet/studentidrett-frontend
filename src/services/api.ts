import { cardType } from '../constants';
import { urlBuilderSimpleSearch } from './urlBuilders';

export const simpleSearch = async (queryUrl: string, cardType: cardType) => {
    try {
        const res = await fetch(urlBuilderSimpleSearch(cardType, queryUrl), {
            method: 'GET',
            headers: new Headers({
                Accept: 'application/json',
            }),
        });
        if (checkForErrorCodes(res)) {
            return {
                count: 0,
                next: null,
                previous: null,
                results: [],
            };
        } else {
            return res.json();
        }
    } catch (error) {
        return 'Connection error';
    }
};

const checkForErrorCodes = (result: any): boolean => {
    return result.status !== 200;
};
