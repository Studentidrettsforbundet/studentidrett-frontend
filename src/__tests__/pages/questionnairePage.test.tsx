import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import QuestionnairePage from '../../pages/questionnairePage';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { fireEvent, render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { handleQuestionsThunk } from '../../services/api';

const mockStore = configureStore([]);

jest.mock('../../services/api', () => ({
    handleQuestionsThunk: jest.fn(() => Promise.resolve()),
}));

const state = {
    questionnaire: {
        questions: [
            {
                id: 'test_id',
                text: 'Test question',
                left: 'Left item',
                right: 'Right item',
            },
        ],
    },
    thunk: { post_success: false, post_in_progress: false, post_failed_count: 0 },
};

describe('questionnairePage', () => {
    let store: any;
    beforeEach(() => {
        store = mockStore(state);
        store.dispatch = jest.fn();
    });

    it('should render correctly', async () => {
        const tree = renderer.create(
            <Provider store={store}>
                <BrowserRouter>
                    <QuestionnairePage />
                </BrowserRouter>
            </Provider>,
        );

        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should send correct request body with post on submit', async () => {
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <QuestionnairePage />
                </BrowserRouter>
            </Provider>,
        );

        fireEvent.click(screen.getByTestId('test_id-3'));
        await act(async () => {
            fireEvent.click(screen.getByText('Send inn'));
        });

        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(handleQuestionsThunk).toHaveBeenCalledWith(false, [{ answer: '3', qid: 'test_id' }]);
    });
});
