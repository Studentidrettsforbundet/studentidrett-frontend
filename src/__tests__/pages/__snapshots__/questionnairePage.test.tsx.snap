// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`questionnairePage should render correctly 1`] = `
<div
  className="f1ipot1o"
>
  <h1>
    Idrettsvalgomat
  </h1>
  <p>
    <span>
      Velkommen til NSIs idrettsvalgomat!
    </span>
    <br />
    <br />
    Du vil bli stilt noen spørsmål om hvilke egenskaper du foretrekker foran i en idrett. Dersom du velger det midterste punktet betyr det at du ikke foretrekker den ene over den andre.
    <br />
    <br />
    Lykke til!
  </p>
  <form
    onSubmit={[Function]}
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <div>
      <div
        className="f15mgf7o"
      >
        <span
          className="f1sgdujd"
        >
          Test question
        </span>
        <div
          className="fj2azb8"
        >
          <span
            className="fk4g4pc"
          >
            <input
              className="fk4g4pc"
              data-testid="test_id-0"
              id="test_id-0"
              name="test_id"
              onChange={[Function]}
              type="radio"
              value={0}
            />
            <label
              htmlFor="test_id-0"
            >
              <span
                className="controller0"
              />
            </label>
          </span>
          <span
            className="fk4g4pc"
          >
            <input
              className="fk4g4pc"
              data-testid="test_id-1"
              id="test_id-1"
              name="test_id"
              onChange={[Function]}
              type="radio"
              value={1}
            />
            <label
              htmlFor="test_id-1"
            >
              <span
                className="controller1"
              />
            </label>
          </span>
          <span
            className="fk4g4pc"
          >
            <input
              className="fk4g4pc"
              data-testid="test_id-2"
              id="test_id-2"
              name="test_id"
              onChange={[Function]}
              type="radio"
              value={2}
            />
            <label
              htmlFor="test_id-2"
            >
              <span
                className="controller2"
              />
            </label>
          </span>
          <span
            className="fk4g4pc"
          >
            <input
              className="fk4g4pc"
              data-testid="test_id-3"
              id="test_id-3"
              name="test_id"
              onChange={[Function]}
              type="radio"
              value={3}
            />
            <label
              htmlFor="test_id-3"
            >
              <span
                className="controller3"
              />
            </label>
          </span>
          <span
            className="fk4g4pc"
          >
            <input
              className="fk4g4pc"
              data-testid="test_id-4"
              id="test_id-4"
              name="test_id"
              onChange={[Function]}
              type="radio"
              value={4}
            />
            <label
              htmlFor="test_id-4"
            >
              <span
                className="controller4"
              />
            </label>
          </span>
        </div>
        <div
          className="f1otibs3"
        >
          <span>
            Left item
          </span>
          <span>
            Right item
          </span>
        </div>
        
      </div>
    </div>
    <button
      className="f10phmif btn btn-custom"
      disabled={false}
      type="submit"
    >
      Send inn
    </button>
  </form>
</div>
`;
